src/databricks/labs/lsql/structs.py

```diff
@@ -0,0 +1,166 @@
+import dataclasses
+import datetime
+import enum
+import types
+from dataclasses import dataclass
+from typing import ClassVar, Protocol, get_args, get_type_hints
+
+
+class StructInferError(TypeError):
+    pass
+
+
+class SqlType(Protocol):
+    def as_sql(self) -> str: ...
+
+
+@dataclass
+class NullableType(SqlType):
+    inner_type: SqlType
+
+    def as_sql(self) -> str:
+        return self.inner_type.as_sql()
+
+
+@dataclass
+class ArrayType(SqlType):
+    element_type: SqlType
+
+    def as_sql(self) -> str:
+        return f"ARRAY<{self.element_type.as_sql()}>"
+
+
+@dataclass
+class MapType(SqlType):
+    key_type: SqlType
+    value_type: SqlType
+
+    def as_sql(self) -> str:
+        return f"MAP<{self.key_type.as_sql()},{self.value_type.as_sql()}>"
+
+
+@dataclass
+class PrimitiveType(SqlType):
+    name: str
+
+    def as_sql(self) -> str:
+        return self.name
+
+
+@dataclass
+class StructField:
+    name: str
+    type: SqlType
+
+    @property
+    def nullable(self) -> bool:
+        return isinstance(self.type, NullableType)
+
+    def as_sql(self) -> str:
+        return f"{self.name}:{self.type.as_sql()}"
+
+
+@dataclass
+class StructType(SqlType):
+    fields: list[StructField]
+
+    def as_sql(self) -> str:
+        fields = ",".join(f.as_sql() for f in self.fields)
+        return f"STRUCT<{fields}>"
+
+    def as_schema(self) -> str:
+        fields = []
+        for field in self.fields:
+            not_null = "" if field.nullable else " NOT NULL"
+            fields.append(f"{field.name} {field.type.as_sql()}{not_null}")
+        return ", ".join(fields)
+
+
+class StructInference:
+    _PRIMITIVES: ClassVar[dict[type, str]] = {
+        str: "STRING",
+        int: "LONG",
+        bool: "BOOLEAN",
+        float: "FLOAT",
+        datetime.date: "DATE",
+        datetime.datetime: "TIMESTAMP",
+    }
+
+    def as_ddl(self, type_ref: type) -> str:
+        v = self._infer(type_ref, [])
+        return v.as_sql()
+
+    def as_schema(self, type_ref: type) -> str:
+        v = self._infer(type_ref, [])
+        if hasattr(v, "as_schema"):
+            return v.as_schema()
+        raise StructInferError(f"Cannot generate schema for {type_ref}")
+
+    def _infer(self, type_ref: type, path: list[str]) -> SqlType:
+        if dataclasses.is_dataclass(type_ref):
+            return self._infer_struct(type_ref, path)
+        if isinstance(type_ref, enum.EnumMeta):
+            return self._infer_primitive(str, path)
+        if type_ref in self._PRIMITIVES:
+            return self._infer_primitive(type_ref, path)
+        if type_ref is list:
+            raise StructInferError("Cannot determine element type of list. Rewrite as: list[XXX]")
+        if type_ref is set:
+            raise StructInferError("Cannot determine element type of set. Rewrite as: set[XXX]")
```

Comment:
I don't think the rest of the code supports sets?

Reply:
nope.

Reply:
apparently, we do support them - `(set[str], "ARRAY<STRING>"),` passed without any modifications